packages/flutter/lib/src/cupertino/dialog.dart

```diff
@@ -196,29 +190,297 @@ class CupertinoAlertDialog extends StatelessWidget {
+        _content = child;
+      } else if (parentData.id == _AlertDialogSection.actions) {
+        _actions = child;
+      }
+    });
+    assert(_content != null);
+    assert(_actions != null);
+
+    final double contentHeight = _content.getMaxIntrinsicHeight(width);
+    final double actionsHeight = _actions.getMaxIntrinsicHeight(width);
+    final double height = min(contentHeight, actionsHeight);
+
+    if (height.isFinite)
+      return height;
+    return 0.0;
+  }
+
+  @override
+  void performLayout() {
+    RenderBox _content;
+    RenderBox _actions;
+    getChildrenAsList().forEach((RenderBox child) {
+      CupertinoDialogRenderBoxParentData parentData = child.parentData;
+      if (parentData.id == _AlertDialogSection.content) {
+        _content = child;
+      } else if (parentData.id == _AlertDialogSection.actions) {
+        _actions = child;
+      }
+    });
+    assert(_content != null);
+    assert(_actions != null);
+
+    // TODO(mattcarroll): parameterize actions count.
+    final actionsCount = 3;
+
+    double minActionSpace = 0.0;
+    if (actionsCount > 0 && actionsCount <= 2) {
+      minActionSpace = _kButtonHeight;
+    } else {
+      minActionSpace = 1.5 * _kButtonHeight;
+    }
+
+    final size = constraints.constrain(const Size(double.infinity, double.infinity));
+
+    // Size alert dialog content.
+    final Size maxContentSize = new Size(size.width, size.height - minActionSpace);
+    _content.layout(
+      new BoxConstraints.loose(maxContentSize),
+      parentUsesSize: true,
+    );
+    final Size contentSize = _content.size;
+
+    // Size alert dialog actions.
+    final Size maxActionSize = new Size(
+      size.width,
+      size.height - contentSize.height,
+    );
+    _actions.layout(
+      new BoxConstraints.loose(maxActionSize),
+      parentUsesSize: true,
+    );
+    final Size actionsSize = _actions.size;
+
+    // Calculate overall dialog height.
+    final double dialogHeight = contentSize.height + actionsSize.height;
+
+    this.size = new Size(size.width, dialogHeight);
+
+    // Layout the blur, content, and the actions.
+    _content.layout(
+      new BoxConstraints.tight(contentSize),
+    );
+    _actions.layout(
+      new BoxConstraints.tight(actionsSize),
+    );
+    (_actions.parentData as CupertinoDialogRenderBoxParentData).offset = new Offset(0.0, contentSize.height);
+  }
+
+  void paint(PaintingContext context, Offset offset) {
+    defaultPaint(context, offset);
+  }
+
+  bool hitTestChildren(HitTestResult result, { Offset position }) {
+    return defaultHitTestChildren(result, position: position);
+  }
+}
+
+class CupertinoDialogRenderBoxParentData extends ContainerBoxParentData<RenderBox> {
+  /// An object representing the identity of this child.
+  Object id;
+
+  @override
+  String toString() => '${super.toString()}; id=$id';
+}
+
+class CupertinoLayoutId extends ParentDataWidget<CupertinoDialogRenderWidget> {
```

Comment:
Just reused LayoutId instead of redefining it

Reply:
I think when I tried LayoutId something blew up because it wasn't being used within a MultiChildLayout, but I'll try it again.

Reply:
Ah sorry, you're right, LayoutId is `extends ParentDataWidget<CustomMultiChildLayout>` which forces the parent widget to be CustomMultiChildLayout. That's unfortunate :(

Reply:
Just refactor the classes so you have a common base class.

Reply:
@Hixie is there a reason for the restriction in the first place?  Why not allow general use of LayoutId?

Reply:
LayoutId fiddles with the parent data, which will only work if the parent data is the right type.